reframe/core/schedulers/sge.py

```diff
@@ -127,36 +130,40 @@ def poll(self, *jobs):
                     continue
 
                 jobid = job_list.find("JB_job_number").text
-
-                if job_number not in [job.jobid for job in jobs]:
-                    # Not a reframe job.
+                if job_number not in jobs_to_poll:
+                    # Not a reframe job
                     continue
 
                 state = job_list.find("state").text
+                job = jobs_to_poll[job_number]
+                known_jobs.add(job)
 
                 # For the list of known statuses see `man 5 sge_status`
                 # (https://arc.liv.ac.uk/SGE/htmlman/htmlman5/sge_status.html)
                 if state in ['r', 'hr', 't', 'Rr', 'Rt']:
-                    jobinfo[job_number] = 'RUNNING'
+                    job._state = 'RUNNING'
                 elif state in ['qw', 'Rq', 'hqw', 'hRwq']:
-                    jobinfo[job_number] = 'PENDING'
+                    job._state = 'PENDING'
                 elif state in ['s', 'ts', 'S', 'tS', 'T', 'tT', 'Rs',
                                'Rts', 'RS', 'RtS', 'RT', 'RtT']:
-                    jobinfo[job_number] = 'SUSPENDED'
+                    job._state = 'SUSPENDED'
                 elif state in ['Eqw', 'Ehqw', 'EhRqw']:
-                    jobinfo[job_number] = 'ERROR'
+                    job._state = 'ERROR'
                 elif state in ['dr', 'dt', 'dRr', 'dRt', 'ds',
                                'dS', 'dT', 'dRs', 'dRS', 'dRT']:
-                    jobinfo[job_number] = 'DELETING'
+                    job._state = 'DELETING'
                 elif state == 'z':
-                    jobinfo[job_number] = 'COMPLETED'
+                    job._state = 'COMPLETED'
+
+        # Mark any "unknown" job as completed
+        unknown_jobs = set(jobs) - known_jobs
+        for job in unknown_jobs:
+            self.log(f'Job {job.jobid} not known to scheduler, '
+                     f'assuming job completed')
+            job._state = 'COMPLETED'
 
-        for job in jobs:
-            if job.jobid not in jobinfo:
-                self.log(f'Job {job.jobid} not known to scheduler, '
-                         f'assuming job completed')
-                job._state = 'COMPLETED'
-                job._completed = True
-                continue
+    def finished(self, job):
+        if job.exception:
+            raise job.exception
 
-            job._state = jobinfo[job.jobid]
+        return job.state == 'COMPLETED'
```

Comment:
Generally, you don't need to send the `_completed` attribute for an SGE job, as far as I can see. This is an artifact of PBS polling, were we need to wait for the actual job output to be written back to the working directory. That's why I redefine the `finished` condition here.